### DISCUSSION TITLE
Regular (automated?) releases and PlantUML lib updates

### DISCUSSION THREAD

**Discussion by travkin79** [hallvard/plantuml#166] [Ideas]
It would be great if we had PlantUML Eclipse plug-in releases soon after new [PlantUML lib](https://github.com/plantuml/plantuml) version releases (and after new Eclipse IDE releases). I'd like to discuss options to reach that goal and clarify how PlantUML Eclipse plug-in releases are done today.

Maybe we can automate all or some of the current build and release steps.

This topic is related to issue #145.

@hallvard describes the current release steps as follows:

> The (manual) process I use to update to a newer plantuml.jar is as follows:
> * acquire the EPL-ed version of the jar and drop it into the libs folder of the corresponding plugin project
> * update all references (to the previous jar) so they refer to the new one
> * update the version of the plugin and feature projects to match the library's and update the reference in the site project
> * build to create the packaged plugin and feature jars
> * metadata about the new library feature must be added to the site composite and it must be pushed to the update site
> All steps can be automated, but it will take time to make it right. You may be able to simplify it, e.g. by renaming the library jar to a standard name. Another approach is hosting the composite library site at plantuml.com and integrate this into their build process.

I have a few first questions concerning building and releasing the PlantUML Eclipse plug-in:

- ~Does the PlantUML lib has to be licensed unter EPL or could we use the GPL-licensed version, too?~
- ~In case, we need the EPL-licensed PlantUML lib:~ Why is it not part of the PlantUML releases? (That might become a new feature request in the [plantuml project](https://github.com/plantuml/plantuml))
- ~I checked out the latest sources from master branch, but they do not compile. Am I doing something wrong?~
  - ~Problem 1: net.sourceforge.plantuml.uml2.Uml2ClassDiagramTextProvider cannot find net.sourceforge.plantuml.ecore.AbstractEcoreClassDiagramTextProvider (it seems not to exist in the sources).~
  - ~Problem 2: net.sourceforge.plantuml.jdt.ClassFileEditorDiagramTextProvider tries to access org.eclipse.jdt.internal.ui.javaeditor.IClassFileEditorInput which results in access restriction errors. Could there be another way to retrieve an IClassFile here?~
- ~I see this plug-in's version 1.1.27 being the latest release according to the update site and according to the README.md, but there seems to be a git tag missing and GitHub shows version 1.1.26 as being the latest release. What can be done to fix that?~
- ~The target platform definition file uses Eclipse Luna and should be updated.~ Can this be automated?
- How could the release steps be automated?
   - ~If the EPL-ed plantuml.jar were available on [Maven Central](https://central.sonatype.com/namespace/net.sourceforge.plantuml) 
      and [mvnrepository](https://mvnrepository.com/), I had some ideas how to automate the first (or the first two) of the steps.~
   - Could we trigger some job as soon as a new PlantUML lib is released (without having to integrate the PlantUML Eclipse plug-in's build process into the build process of the PlantUML lib)?

*Update*
In the meanwhile, I solved the crossed out questions.

**Comment by arnaudroques** [comment]
Hello,

Thanks for starting this discussion.
The fact is that we are looking for a new maintainer for the Eclipse plugin :-)
@hallvard did a great job on this project but he cannot manage it anymore.

So if someone is volunteer (@travkin79 it could be you!), please tell us.
 
<hr>

> If the EPL-ed plantuml.jar were available on [mvnrepository](https://mvnrepository.com/), I had some ideas how to automated the first (or the first two) of the steps.

The good news is that [we are working on it](https://github.com/plantuml/plantuml/discussions/1336)!

**Comment by travkin79** [comment]
Hello @arnaudroques,
Thanks for asking. I'm not sure if can manage to maintain the plug-in, but I could give it a try to prepare a new release, probably initially "by hand" and then try to simplify / automate the release process.

Is there someone, who could help me answering some of the questions?
**Update**
I solved the two compile problems mentioned above. It turns out the uml2 plug-in is no longer maintained and I could avoid the access restriction by slightly adapting the code. Details here: https://github.com/travkin79/plantuml4eclipse

Nice to see that you're already working on releasing EPL-ed plantuml.jar files and support other licenses, too. 👍

**Reply by arnaudroques** [reply]
> Thanks for asking. I'm not sure if can manage to maintain the plug-in, but I could give it a try to prepare a new release, probably initially "by hand" and then try to simplify / automate the release process.
> 

Ok, let's go step by step.
To start, you can try to prepare a new release. That's really a very good first step.

> Is there someone, who could help me answering some of the questions?

Yes, I think that @hallvard could help you to answer some of your questions.
And this is open source, the community may help you also. 

> **Update** I solved the two compile problems mentioned above. It turns out the uml2 plug-in is no longer maintained and I could avoid the access restriction by slightly adapting the code. Details here: https://github.com/travkin79/plantuml4eclipse

Great, to me, it means that you are ready :-)
If you are ok, I'll ask @hallvard to add you direct write access to the repository.
You can also email us at plantuml@gmail.com if you have other questions.

**Comment by travkin79** [comment]
Hello @arnaudroques and @hallvard,
Thank you. I've created a fork, but if you give me write access, things might get easier. :-D

**Comment by travkin79** [comment]
Thanks for the access, @hallvard.

I've prepared a new PlantUML Eclipse plug-in release candidate in my fork ([https://github.com/travkin79/plantuml4eclipse/tree/feature/145](https://github.com/travkin79/plantuml4eclipse)) with the following changes.
* Set next version to 1.1.28-SNAPSHOT
* Update PlantUML lib version to 1.2023.4 (EPL version)
* Add target platform definition for Eclipse 2022-09
* Adapt Maven build to use the new target definition file (instead of using p2 repositories and being dependent from the local system)
* Switch to using Java 11 as minimum requirement (because Eclipse 2022-09 needs Java 11)
* Fix most of the tests, incl. UI tests. But I could not fix `TextEditorDiagramTextProviderTest2`.

Could you take a look at the changes, before we're taking next steps for merging the changes to master and preparing a 1.1.28 release?

**Comment by travkin79** [comment]
I updated my implementation to PlantUML lib version 1.2023.5 and created a pull request #169 for review purposes.

**Comment by travkin79** [comment]
Hi @arnaudroques,
As @hallvard already described, we need to wrap the PlantUML library in an OSGi / Eclipse bundle and provide an Eclipse update site in order to make the PlantUML library re-usable in Eclipse plug-ins. I'd like to automatically create such bundles with each PlantUML release and use that bundle in the PlantUML Eclipse plug-ins as well as in some other Eclipse plug-ins. Anyone else using PlantUML in Eclipse plug-ins could do the same.

https://github.com/hallvard/plantuml/issues/145#issuecomment-976247581:
>..., it [the library] needs to be wrapped in an Eclipse plugin and published on the update site. This should ideally be automated, but that requires an effort I haven't found time for.

There is a Gradle plug-in that could be used to automate that step and solve a major part of issue #145:
https://github.com/stempler/bnd-platform

I would give it a try and prepare a patch for the [plantuml/plantuml](https://github.com/plantuml/plantuml) project. But I have a few questions left:

- Is it ok, to extend the build steps for PlantUML EPL version to automatically bundle the plantuml-epl.jar file in an OSGi bundle and create an eclipse update site?
- Can we (automatically) publish the update site with each PlantUML release on GitHub pages (or somewhere else)? The update site (sometimes called p2 repository) is just a directory tree with some files that have to be accessible via URL in order to be used in Eclipse plug-ins.
- Can the publishing steps in GitHub workflow files be tested in a fork?

**Reply by arnaudroques** [reply]
>  * Is it ok, to extend the build steps for PlantUML EPL version to automatically bundle the plantuml-epl.jar file in an OSGi bundle and create an eclipse update site?
> 

Yes, it's ok for me. @The-Lum is it ok for you also? 

>  * Can we (automatically) publish the update site with each PlantUML release on GitHub pages (or somewhere else)? The update site (sometimes called p2 repository) is just a directory tree with some files that have to be accessible via URL in order to be used in Eclipse plug-ins.
>

Yes, we could host this on PlantUML GitHub pages


>  *  Can the publishing steps in GitHub workflow files be tested in a fork?
>

Not sure if we could test this in a fork or a branch: but we could make a try :-)
You can propose PR on the three subjects!

**Reply by The-Lum** [reply]
>Yes, it's ok for me. @The-Lum is it ok for you also?

For me that is ok. _And if the users of the EPL plantuml version also._

**Comment by travkin79** [comment]
For traceability:
After some experiments in a PR https://github.com/plantuml/plantuml/pull/2127, we go on discussing this topic in a new git repository here: https://github.com/plantuml/plantuml-eclipse/discussions/2.
